Add actions to build PRs and develop, cleanup closed PRs.

adamsilverstein opened the pull request:
## Summary
This PR adds two [GitHub Actions](https://github.com/features/actions) to automatically build (and later cleanup) versions of the plugin for testing. These build files can be used to set the version of Site Kit to any PR or develop for testing.

 * For each PR that is opened or updated, the `build-pr` action builds a zip production (`google-site-kit.zip`) and development(`google-site-kit-dev.zip`) build version of the plugin.
* The zip files are moved to a folder based on the current [GitHub Ref ](https://developer.github.com/v3/git/refs/). In the case of pull requests this is `refs/pull/[PR#]/merge`
* The action also builds when PRs are merged to the `develop` branch, in this case the ref is `refs/heads/develop`. 
* The zip files/ref folder is committed to the repository's wiki repository.
* These folders create predictable zip URLs to download the correct build for testing.
* When a pull request is closed the `remove-pr` action runs and the ref folder and zip files are deleted from the wiki repository.

## Relevant technical choices
* The build process is cached so only the initial run needs to download `npm` and `composer` assets (unless they change)
* The build uses several npm/gulp commands to build the desired plugin zips. These could be combined into a single npm command. My goal in adding them this way is to add these actions without changing the plugin at all.
* For upload locations, I considered several other solutions including committing to a branch, adding PR comments, using releases or releases on a separate repository and using build artifacts. The wiki solution wound up being the best because no extra repository is required, the primary repository is not weighed down by constantly added zip files, and the built files are transparent to end users.

adamsilverstein commented:
@felixarntz someone with repo ownership will need to add a personal access token as a repo secret under the key `GITHUB_PERSONAL_ACCESS_TOKEN` before the wiki deploys can work correctly. 

Also I noticed the build failed, possible because of some recent changes in build commands? I'll take a look and get that fixed.

adamsilverstein commented:
Fixed the build command, I had to adjust for the newer `grunt build` command which creates a unique zip filename when building. 

Once the `GITHUB_PERSONAL_ACCESS_TOKEN` is added to site secrets, the built files will get added to the wiki repo.

felixarntz commented on .github/workflows/build-pr.yml:
```diff
@@ -0,0 +1,60 @@
+        npm install gulp-cli -g
+    - name: Build develop version
+      run: |
+        npm run build:dev
```
Not really scope of this PR, but the scripts in `package.json` are a bit inconsistent between how they work for production vs development. E.g. `build:dev` includes a call to `build:static` while `build:production` and `build:test` don't.

I'll open a small separate issue with this.

aaemnnosttv commented on .github/release-to-wiki.sh:
```diff
@@ -0,0 +1,28 @@
+mkdir -p "$tmp_dir/$1"
+for file in $(find $1 -maxdepth 1 -type f -name '*' -execdir basename '{}' ';'); do
+    cp "$1/$file" "$tmp_dir/$1"
+done
```
Why is this necessary? Why is it only copying top-level files?

adamsilverstein replied:
it really only needs to copy the two zip files, I'll just make it two explicit copies to make it clearer and less generic.

aaemnnosttv replied:
Ah, in that case you could just use a simple glob: `cp dir/*.zip dest-dir` 👍

adamsilverstein replied:
Nice! Updated.

adamsilverstein commented on .github/workflows/build-pr.yml:
```diff
@@ -0,0 +1,60 @@
+      run: |
+        npm run build
+        ZIP=`gulp release | grep Creating | cut -c10-`
+        mv "$ZIP" google-site-kit.zip
```
Good suggestion, I had thought about and put that off. I'll clean this up.

adamsilverstein replied:
updated this and added a new `dev-zip` npm command.

adamsilverstein commented on .github/workflows/build-pr.yml:
```diff
@@ -0,0 +1,60 @@
+    - name: Setup Node.js 10
+      uses: actions/setup-node@v1
+      with:
+        node-version: '10.x'
```
Oh, cool! will update.

adamsilverstein replied:
Updated in https://github.com/google/site-kit-wp/pull/1172/commits/fcd7b1c6c11a6263c03d611be4864eac00aa3a77

adamsilverstein commented on .github/workflows/build-pr.yml:
```diff
@@ -0,0 +1,60 @@
+    - name: npm/Gulp Install
+      run: |
+        npm install
+        npm install gulp-cli -g
```
ok, will change. I did't even realize the project included gulp, locally I think I've always relied on the global install.

adamsilverstein replied:
removed, no longer using gulp directly anyway.

adamsilverstein commented on .github/workflows/build-pr.yml:
```diff
@@ -0,0 +1,60 @@
+  push:
+    branches:
+      - develop
+  pull_request:
```
Yes pushes to the PR initiate a rebuild. By [default](https://help.github.com/en/actions/reference/events-that-trigger-workflows#pull-request-event-pull_request) (with no types passed), a workflow runs when a pull_request's activity type is **opened**, **synchronize**, or **reopened**.

adamsilverstein commented on .github/release-to-wiki.sh:
```diff
@@ -0,0 +1,28 @@
+    git pull "$GIT_REPOSITORY_URL"
+)
+
+mkdir -p "$tmp_dir/$1"
```
but... this is how we do bash :) will change.

adamsilverstein replied:
adjusted: moved to passing GIT_REF in env (in action)

adamsilverstein commented on .github/release-to-wiki.sh:
```diff
@@ -0,0 +1,28 @@
+
+GIT_REPOSITORY_URL="https://${GITHUB_PERSONAL_ACCESS_TOKEN}@github.com/$GITHUB_REPOSITORY.wiki.git"
+
+tmp_dir=$(mktemp -d -t ci-XXXXXXXXXX)
```
good suggestion, I'll change it.

adamsilverstein replied:
updated.

adamsilverstein commented on .github/release-to-wiki.sh:
```diff
@@ -0,0 +1,28 @@
+    git init
+    git config user.name "$GITHUB_ACTOR"
+    git config user.email "$GITHUB_ACTOR@users.noreply.github.com"
+    git pull "$GIT_REPOSITORY_URL"
```
hmm, would that be faster? I'll try changing.

aaemnnosttv replied:
I don't know if it would be _faster_, but it wouldn't require the init command (no code _is_ faster than no code 😉 ). I've never seen it done this way before so partially a bit curious here myself 😄

adamsilverstein replied:
Possibly done this way so that it also works for private repos (since you add credentials before pulling)? I agree its cleaner to use clone so I changed it.

aaemnnosttv replied:
Maybe? 🤔 I don't think user name/email are used for authentication though.

adamsilverstein commented:
@aaemnnosttv Thanks for the review, I'll work on addressing your feedback

> I actually created a Docker container recently for running the plugin build in a reproducible way. I'm not sure if that could be used here (it only supports a production build and a specific branch but also supports using a local repo) but might be useful to see if anything can be borrowed there. [aaemnnosttv/gskbuild](https://github.com/aaemnnosttv/gskbuild)

Very cool! I have a separate tester plugin I am working on to leverage these zip builds, more on that soon!

adamsilverstein commented:
@aaemnnosttv This is ready for re-review, I have addressed your feedback and cleaned things up a bit. Thanks especially for the bash script improvements. 

We added the required secret and ZIPs are building now (for this pr only so far), you can clone the wiki repo to see them: `git clone https://github.com/google/site-kit-wp.wiki.git`

aaemnnosttv commented on .github/workflows/build-pr.yml:
```diff
@@ -0,0 +1,64 @@
+        mkdir -p ${{ github.ref }}
+        mv google-site-kit.zip ${{ github.ref }}
+        mv google-site-kit-dev.zip ${{ github.ref }}
+        bash .github/release-to-wiki.sh
```
No reason to rename and move the zips in separate steps since this can be done in the same command 😄 

```suggestion
    - name: Create destination directory
      run: mkdir -p ${{ github.ref }}
    - name: Build develop version
      run: |
        npm run dev-zip
        mv *.zip ${{ github.ref }}/google-site-kit-dev.zip
    - name: Build release version
      run: |
        npm run release-zip
        mv *.zip ${{ github.ref }}/google-site-kit.zip
    - name: Deploy build files to wiki
      run: bash .github/release-to-wiki.sh
```

adamsilverstein replied:
Ah great! I should have read this first then I could have applied it before changing myself. Should match your proposal now though, please review :)

adamsilverstein commented on .github/workflows/build-pr.yml:
```diff
@@ -42,13 +42,12 @@ jobs:
       run: |
         npm run build
-        ZIP=`gulp release | grep Creating | cut -c10-`
+        ZIP=`npm run dev-zip | grep Creating | cut -c10-`
```
Swapped them.

Ok, I'll change to that mv, that is smart though not as clever as my solution.

adamsilverstein replied:
changed. this works as long as we don't add zips to the project root :)

aaemnnosttv replied:
I think we're safe 😄

adamsilverstein commented on package.json:
```diff
@@ -10,6 +10,7 @@
     "build:dev": "npm run build:static && webpack --mode=development --debug --devtool cheap-source-map --output-pathinfo",
     "build:static": "gulp svg imagemin",
     "dev": "npm run build:dev && npm run build:static",
+    "dev-zip": "npm run build:dev && gulp release",
```
hmmm... wouldn't I add `npm run remove-dist` to my `dev-zip` command? not following I guess, can you commit the fix you propose directly.

aaemnnosttv replied:
We also want the dist directory to be cleared when doing a dev build, not just when creating a zip for one 😉 

As Felix said before, these commands need cleaning up for consistency, but I'll make a temporary fix for now.

adamsilverstein replied:
👍 right, got it! it cleans up the dist folder, not the zip, I was mixing that up.

adamsilverstein commented:
@aaemnnosttv Ready for another review, except [this bit](https://github.com/google/site-kit-wp/pull/1172#discussion_r385893691) which I don't quite understand, can you fix that or explain further?

aaemnnosttv commented:
> @aaemnnosttv Ready for another review, except this bit which I don't quite understand, can you fix that or explain further?

Fixed! See explanation [here](https://github.com/google/site-kit-wp/pull/1172#discussion_r385911722)

There are a few more enhancements I'd like to make that I think can really simplify things but I'm out of time today. I'll look at this again tomorrow or Monday.

adamsilverstein commented:
Thanks!

aaemnnosttv commented:
@googlebot I consent.

aaemnnosttv commented:
@adamsilverstein I've made a few updates to the workflow to simplify things a bit which I think you will like!

* Completely defined in workflow files, no need for external bash scripts
* Using `actions/checkout` there is no need for an additional access token as an [auth token is persisted in the repos local git config by default](https://github.com/actions/checkout#checkout-v2)
* I split zip/build and deploy tasks into separate jobs to avoid potential conflicts with the working directory between repos using artifacts upload+download
* Using artifacts makes the zips available for download from the Actions backend for each run which are available for 90 days  
![image](https://user-images.githubusercontent.com/1621608/75606389-dfb62e00-5af4-11ea-9976-fb085bbd1c79.png)
* I set the git config name and email with environment variables instead of git commands which makes that portion of the script a bit cleaner and focused I think  
**Should instead be something static and generic like `googlebot` rather than the GitHub actor since its done via automation?**

Let me know what you think or if there's anything I might have missed.

aaemnnosttv commented:
Requesting @tofumatt's eyes here since he is quite familiar with Actions as well 👍

adamsilverstein commented:
@aaemnnosttv Thanks for the refactoring here! 

> I split zip/build and deploy tasks into separate jobs to avoid potential conflicts with the working directory between repos using artifacts upload+download

Nice.

> Using artifacts makes the zips available for download from the Actions backend for each run which are available for 90 days

Did you figure out a way to determine or specify an artifact's URL via API or after creation? If we knew this we could use build artifacts instead of the wiki.

> Completely defined in workflow files, no need for external bash scripts

Excellent, that is much cleaner and easier to understand as well!

> there is no need for an additional access token

Great! I'll remove my token.

> Should instead be something static and generic like googlebot rather than the GitHub actor since its done via automation?

the github actor feels clear enough to me.

aaemnnosttv commented:
> Did you figure out a way to determine or specify an artifact's URL via API or after creation? If we knew this we could use build artifacts instead of the wiki.

I thought about that but the fact that they expire is a bit of a dealbreaker - also, they're specific to the individual action run, so I don't think that there is an easy way to grab the latest like we are doing with the URL to the wiki. Also, build artifacts are put into a directory even if you upload an individual file - the artifact comes down as a zipped directory. That is, you wouldn't be able to link to an artifact of the installable zip, unless you used `upload-artifact` to target the unzipped release zip (does that make sense?).  Because of this, both zips are archived together inside the destination directory, and then restored back into it 🙂

tofumatt commented on .github/workflows/build-pr.yml:
```diff
@@ -0,0 +1,83 @@
+name: build-pr
+
+on:
```
If we build the PR on any tags/and on `master` it would save us (usually @felixarntz) having to build the production ZIP by hand on a Mac. Having the CI generate our builds instead of us doing it manually is way nicer, so I filed #1196 to chat about that and extend this later 😄

aaemnnosttv replied:
Good call!

tofumatt commented on .github/workflows/build-pr.yml:
```diff
@@ -0,0 +1,83 @@
+      uses: php-actions/composer@v1
+      with:
+        args: install
+    - name: Read .nvmrc
```
That's a nice touch. I usually just specify the node version in the action, but I like that! 👍

tofumatt commented on .github/workflows/build-pr.yml:
```diff
@@ -0,0 +1,83 @@
+on:
+  push:
+    branches:
+      - develop
```
This runs on pushes to `develop`, but doesn't remove the file. Will the upload overwrite this one based on the `ref` being the branch name (`develop`) over the commit hash?

If so that seems fine, but just wanna make sure this isn't going to spawn a lot of ZIP files 😅

aaemnnosttv replied:
The generated zip files are renamed as `google-site-kit.zip` and `google-site-kit-dev.zip` so yes 😄

adamsilverstein replied:
note: the github.ref in this case is `refs/heads/develop` (so that is where the zips wind up being stored)

adamsilverstein commented on .github/workflows/build-pr.yml:
```diff
@@ -0,0 +1,83 @@
+        uses: actions/download-artifact@v1
+        with:
+          name: zip-files
+          path: ${{ github.ref }}
```
Love how this works!

adamsilverstein commented:
Excellent updates @aaemnnosttv - thank you! +1
